### PR TITLE
Curriculum feature

### PR DESCRIPTION
**Pull request by Shirosaya25**
Renamed CurrModules to CurriculumModules
Working on writing tests - told the others to commit to this branch

### CODE REVIEW

File: src/main/java/com/revature/controllers/CurriculumModuleController.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.revature.entities.CurriculumModule;
+import com.revature.services.CurriculumModuleService;
+
+@CrossOrigin(origins = "*", allowCredentials="true")
+@RestController
+@RequestMapping("/curriculummodules")
```
Comment (carolynrehm) [inline src/main/java/com/revature/controllers/CurriculumModuleController.java]: curriculum-modules would be the appropriate multi-word restful convention

File: src/main/java/com/revature/entities/CurriculumModule.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@Entity
+@Table(name="currmodule")
```
Comment (carolynrehm) [inline src/main/java/com/revature/entities/CurriculumModule.java]: Extend these names: curriculum_module, curriculum_module_id, etc

File: src/main/java/com/revature/services/CurriculumModuleServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+import com.revature.entities.CurriculumModule;
+import com.revature.repositories.CurriculumModuleRepository;
+import com.revature.util.LogException;
+
+@Service
+public class CurriculumModuleServiceImpl implements CurriculumModuleService  {
+
+	@Autowired
+	CurriculumModuleRepository cmR;
```
Comment (carolynrehm) [inline src/main/java/com/revature/services/CurriculumModuleServiceImpl.java]: make these variables more readable, its okay if they're long (cmR, currM, etc)

Comment (carolynrehm) [reply src/main/java/com/revature/services/CurriculumModuleServiceImpl.java]: +CurriculumServiceImpl

File: src/main/java/com/revature/util/MetricsData.java [hunk 4]
```diff
@@ -1,109 +1,109 @@
...
-		return true;
-	}
-	
-	
-	
-	
-}
+package com.revature.util;
+
+public class MetricsData { // Bean for the metrics
```
Comment (carolynrehm) [inline src/main/java/com/revature/util/MetricsData.java:3]: This file does not seem to be changed, exclude it if unnecessary

File: src/test/java/com/revature/controllertests/CurriculumControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.revature.controllertests;
+
+import org.springframework.boot.test.context.SpringBootTest;
+
+import com.revature.cmsforce.CMSforceApplication;
+
+@SpringBootTest(classes = CMSforceApplication.class)
```
Comment (carolynrehm) [inline src/test/java/com/revature/controllertests/CurriculumControllerTest.java]: I'd recommend excluding this and creating a branch off for testing - that way we can get these changes in sooner

File: src/main/java/com/revature/controllers/CurriculumController.java [hunk 6]
```diff
@@ -61,8 +61,8 @@
 		return ResponseEntity.ok(curriculumService.createCurriculum(curriculum));
 	}
 
-	@PutMapping("/{id}")
-	public ResponseEntity<Curriculum> updateCurriculum(@PathVariable("id") int id, @RequestBody Curriculum curriculum) {
+	@PutMapping
```
Comment (carolynrehm) [inline src/main/java/com/revature/controllers/CurriculumController.java]: Why was the id param removed here? It's convention to have it in the url

**Review by carolynrehm — CHANGES_REQUESTED** [review]
See previous comments for changes. 

I would create a new branch for future tests; this will allow you to get this feature merged in both here and on the front end

**Comment by carolynrehm** [comment]
Can we also get a short script to convert the current database to the new data model?

File: src/test/java/com/revature/controllertests/CurriculumModuleControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        ResultActions result = mvc.perform( post("/curriculummodules")
+                .contentType(MediaType.APPLICATION_JSON_VALUE)
+                .content(objMapper.writeValueAsString(curriculumModule)));
+
+        CurriculumModule actual = objMapper.readValue(result.andReturn().getResponse()
+                .getContentAsString(), CurriculumModule.class);
+
+        result.andExpect(status().isOk());
+        Mockito.verify(cms).createCurrModule(curriculumModule);
+        Assert.assertEquals(actual, curriculumModule, "Failed to retrieve expected curriculum module");
```
Comment (carolynrehm) [inline src/test/java/com/revature/controllertests/CurriculumModuleControllerTest.java]: Each test should have only one thing that could make it fail (expect ok  / assert method)

Comment (carolynrehm) [reply src/test/java/com/revature/controllertests/CurriculumModuleControllerTest.java]: +GetAllCurriculumModuleById +GetAllCurriculumModules

File: src/main/java/com/revature/controllers/CurriculumController.java (hunk 6 above)
Comment (reimerkp) [reply src/main/java/com/revature/controllers/CurriculumController.java]: we weren't actually using the id param but I can replace it

Comment (AdamRev) [reply src/main/java/com/revature/controllers/CurriculumController.java]: You should use the id in the path to ensure the user is updating the correct record :)

File: src/main/java/com/revature/services/CurriculumModuleServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+
+	@Override
+	public Iterable<CurriculumModule> updateCurrModule(Set<CurriculumModule> curriculumModule) {
+		return curriculumModuleRepository.saveAll(curriculumModule);
+	}
+
+	@Override
+	public void deleteCurriculumModule(CurriculumModule curriculumModule) {
```
Comment (carolynrehm) [inline src/main/java/com/revature/services/CurriculumModuleServiceImpl.java:46]: make these method signatures consistent

File: src/main/java/com/revature/controllers/CurriculumModuleController.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+		return ResponseEntity.ok(currMods);
+	}
+	
+	@GetMapping(value="/{id}")
+	public ResponseEntity<CurriculumModule> getCurrModuleById(@PathVariable int id){
+		CurriculumModule currMod = currModuleService.getCurrModuleById(id);
+		return ResponseEntity.ok(currMod);
+	}
+	
+	@PutMapping("/update")
```
Comment (carolynrehm) [inline src/main/java/com/revature/controllers/CurriculumModuleController.java]: y

Comment (reimerkp) [reply src/main/java/com/revature/controllers/CurriculumModuleController.java]: I don't know what you want me to do with this information

Comment (carolynrehm) [reply src/main/java/com/revature/controllers/CurriculumModuleController.java]: "update" for a post request does not follow restful conventions! Just mentioned in CurriculumController comments. https://restfulapi.net can clarify naming conventions if they're still unclear

**Review by carolynrehm — COMMENTED** [review]
I think the plural of curriculum would be more appropriate as /curricula but curriculum would probably be fine. For the put mappings, you'd want to set the current entities id to the id in the parameter to assure you're updating the correct record rather than overwriting to be a different record (updateCurriculum and updateCurriculumModule). Also, for updateCurriculumModule, you were right when you had set the path to just "{id}".

Resolve these and the comment on testing and this branch should be ready to merge in

**Comment by reimerkp** [comment]
All completed tests are in a separate branch called curriculum_feature_tests

File: src/test/java/com/revature/controllertests/CurriculumModuleControllerTest.java (hunk 7 above)
Comment (carolynrehm) [reply src/test/java/com/revature/controllertests/CurriculumModuleControllerTest.java]: moved to new branch  curriculum_feature_tests

**Review by carolynrehm — APPROVED** [review]

**Pull request merged**